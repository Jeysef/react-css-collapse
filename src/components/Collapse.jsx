import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import useCollapse from './useCollapse';

const Collapse = ({
  children,
  className,
  isOpen,
  transition,
  onRest,
  ...attrs
}) => {
  const content = useRef(null);
  const { expand, collapse, style } = useCollapse({ transition, isOpen, content });

  const onTransitionEnd = (e) => {
    if (e.target === content && e.propertyName === 'height') {
      if (isOpen) {
        expand();
      } else {
        collapse();
      }
      if (onRest) {
        onRest();
      }
    }
  };

  return (
    <div
      ref={content}
      style={style}
      className={className}
      onTransitionEnd={onTransitionEnd}
      {...attrs}
    >
      {children}
    </div>
  );
};

Collapse.displayName = 'Collapse';

Collapse.defaultProps = {
  children: null,
  className: 'react-css-collapse-transition',
  isOpen: false,
  transition: null,
  onRest: null,
};

Collapse.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  isOpen: PropTypes.bool,
  transition: PropTypes.string,
  onRest: PropTypes.func,
};

export default Collapse;
