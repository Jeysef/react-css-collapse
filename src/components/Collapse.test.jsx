import React from 'react';
import { render, act } from '@testing-library/react';
import Collapse from './Collapse';

describe('Collapse', () => {
  test('default', async () => {
    let component;
    await act(async () => {
      component = render(<Collapse />);
    });
    expect(component.asFragment()).toMatchSnapshot();
  });

  test('with children', async () => {
    let component;
    await act(async () => {
      component = render(<Collapse>children</Collapse>);
    });
    expect(component.asFragment()).toMatchSnapshot();
  });

  test('with class', async () => {
    let component;
    await act(async () => {
      component = render(<Collapse className="className" />);
    });
    expect(component.asFragment()).toMatchSnapshot();
  });

  test('with style', async () => {
    let component;
    await act(async () => {
      component = render(
        <Collapse
          style={{ transition: 'height 1337ms cubic-bezier(.4, 0, .2, 1)' }}
        />,
      );
    });
    expect(component.asFragment()).toMatchSnapshot();
  });

  test('with transition prop', async () => {
    let component;
    await act(async () => {
      component = render(
        <Collapse transition="height 1337ms cubic-bezier(.4, 0, .2, 1)" />,
      );
    });
    expect(component.asFragment()).toMatchSnapshot();
  });

  test('with aria and data props', async () => {
    let component;
    await act(async () => {
      component = render(<Collapse aria-hidden data-any-value="1337" />);
    });
    expect(component.asFragment()).toMatchSnapshot();
  });

  test('when open', async () => {
    let component;
    await act(async () => {
      component = render(<Collapse isOpen />);
    });
    expect(component.asFragment()).toMatchSnapshot();
  });
});
