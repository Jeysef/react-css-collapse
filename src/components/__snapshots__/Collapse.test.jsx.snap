// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse default 1`] = `
<DocumentFragment>
  <div
    class="react-css-collapse-transition"
    style="will-change: height; overflow: hidden; visibility: hidden; height: 0px;"
  />
</DocumentFragment>
`;

exports[`Collapse when open 1`] = `
<DocumentFragment>
  <div
    class="react-css-collapse-transition"
    style="will-change: height; overflow: hidden; visibility: visible; height: auto;"
  />
</DocumentFragment>
`;

exports[`Collapse with aria and data props 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="react-css-collapse-transition"
    data-any-value="1337"
    style="will-change: height; overflow: hidden; visibility: hidden; height: 0px;"
  />
</DocumentFragment>
`;

exports[`Collapse with children 1`] = `
<DocumentFragment>
  <div
    class="react-css-collapse-transition"
    style="will-change: height; overflow: hidden; visibility: hidden; height: 0px;"
  >
    children
  </div>
</DocumentFragment>
`;

exports[`Collapse with class 1`] = `
<DocumentFragment>
  <div
    class="className"
    style="will-change: height; overflow: hidden; visibility: hidden; height: 0px;"
  />
</DocumentFragment>
`;

exports[`Collapse with style 1`] = `
<DocumentFragment>
  <div
    class="react-css-collapse-transition"
    style="will-change: height; transition: height 1337ms cubic-bezier(.4, 0, .2, 1); overflow: hidden; visibility: hidden; height: 0px;"
  />
</DocumentFragment>
`;

exports[`Collapse with transition prop 1`] = `
<DocumentFragment>
  <div
    class="react-css-collapse-transition"
    style="will-change: height; transition: height 1337ms cubic-bezier(.4, 0, .2, 1); overflow: hidden; visibility: hidden; height: 0px;"
  />
</DocumentFragment>
`;
